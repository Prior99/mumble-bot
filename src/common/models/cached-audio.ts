import { is, scope, DataType, specify, uuid } from "hyrest";

import { world, enqueue, live, createSound } from "../scopes";

import { User } from ".";

/**
 * A cached audio.
 */
export class CachedAudio {
    constructor(id?: string, user?: User, duration?: number, date?: Date) {
        if (id && user && duration) {
            this.user = user;
            this.duration = duration;
            this.id = id;
        }
        if (date) {
            this.date = date;
        } else {
            this.date = new Date();
        }
    }

    /**
     * The date the audio was recorded.
     */
    @is() @scope(world, live) @specify(() => Date)
    public date?: Date;

    /**
     * The user from which the audio was recorded.
     */
    @is() @scope(world, live)
    public user?: User;

    /**
     * The id of the cached audio.
     */
    @scope(world, enqueue, live, createSound) @is().validate(uuid)
    public id?: string;

    /**
     * The duration of the audio in seconds.
     */
    @is(DataType.float) @scope(world, live)
    public duration?: number;
}
