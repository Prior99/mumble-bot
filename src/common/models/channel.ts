import { is, DataType, specify } from "hyrest";

import { MumbleUser } from ".";

export class Channel {
    @is()
    public name?: string;

    @is(DataType.int)
    public position?: number;

    @is() @specify(() => MumbleUser)
    public users?: MumbleUser[];

    @is() @specify(() => Channel)
    public children?: Channel[];
}
