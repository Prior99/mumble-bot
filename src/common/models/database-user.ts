import { Column, PrimaryGeneratedColumn, Entity, OneToMany } from "typeorm";
import { is, scope, specify, length, uuid, transform, only, required } from "hyrest";

import { world, login, owner, signup } from "../scopes";
import { hash } from "../utils";

import { Recording, PermissionAssociation, Token, Setting, MumbleLink } from ".";

/**
 * A user from the database.
 */
@Entity()
export class DatabaseUser {
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: string;

    /**
     * The username of this user.
     */
    @Column("varchar", { length: 100 })
    @is()
        .validate(length(3, 100), only(signup, required))
        .validateCtx(ctx => only(signup, value => ctx.validation.nameAvailable(value)))
    @scope(world, signup)
    public name?: string;

    /**
     * The password of the user.
     */
    @Column("varchar", { length: 200 })
    @transform(hash)
    @is().validate(length(8, 255)) @scope(login)
    public password?: string;

    @Column("varchar", { length: 200 })
    @is()
        .validate(length(8, 200), required)
        .validateCtx(ctx => only(signup, value => ctx.validation.emailAvailable(value)))
    @scope(login)
    public email?: string;

    @OneToMany(() => Recording, recording => recording.user)
    @is() @scope(world) @specify(() => Recording)
    public recordings?: Recording[];

    @OneToMany(() => Recording, recording => recording.reporter)
    @is() @scope(world) @specify(() => Recording)
    public reported?: Recording[];

    @OneToMany(() => PermissionAssociation, permissionAssociation => permissionAssociation.user)
    @is() @scope(world) @specify(() => PermissionAssociation)
    public permissionAssociations?: PermissionAssociation[];

    @OneToMany(() => Token, token => token.user)
    @is() @specify(() => Token)
    @scope(owner)
    public tokens?: Token[];

    /**
     * The custom settings of the user are stored key-value-wise in this object.
     */
    @OneToMany(() => Setting, setting => setting.user)
    @is() @specify(() => Setting)
    @scope(owner)
    public settings?: Setting[];

    @OneToMany(() => MumbleLink, mumbleLink => mumbleLink.user)
    @is() @specify(() => MumbleLink)
    public mumbleLinks?: MumbleLink;
}
