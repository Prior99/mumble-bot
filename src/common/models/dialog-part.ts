import { PrimaryGeneratedColumn, Entity, OneToMany } from "typeorm";
import { is, scope, uuid } from "hyrest";

import { world, createDialog } from "../scopes";

import { Dialog, Recording  } from ".";

@Entity()
export class DialogPart {
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: string;

    @OneToMany(() => Recording, recording => recording.dialogParts)
    @is() @scope(world, createDialog)
    public recording?: Recording;

    @OneToMany(() => Dialog, dialog => dialog.parts)
    @is() @scope(world)
    public dialog?: Dialog;
}
