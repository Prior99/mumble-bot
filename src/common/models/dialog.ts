import { Column, PrimaryGeneratedColumn, Entity, ManyToOne, CreateDateColumn } from "typeorm";
import { is, scope, specify, uuid, DataType } from "hyrest";
import { world, createDialog } from "../scopes";
import { DialogPart } from ".";

/**
 * A dialog as represented in the database including all its records.
 * @typedef Dialog
 */
@Entity()
export class Dialog {
    /**
     * Unique id of this dialog.
     */
    @PrimaryGeneratedColumn("uuid")
    @scope(world) @is().validate(uuid)
    public id?: number;

    /**
     * The date when this dialog was submitted.
     */
    @CreateDateColumn()
    @scope(world) @is() @specify(() => Date)
    public submitted?: Date;

    /**
     * How often this dialog was used.
     */
    @Column("int", { default: 0 })
    @is(DataType.int) @scope(world)
    public used?: number;

    /**
     * All records belonging to this dialog.
     */
    @ManyToOne(() => DialogPart, dialogPart => dialogPart.dialog)
    @is() @scope(world, createDialog) @specify(() => DialogPart)
    public parts?: DialogPart[];
}
