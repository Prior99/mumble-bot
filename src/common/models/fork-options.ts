import { is, DataType, specify, required, scope } from "hyrest";
import { fork } from "../scopes";

export class Action {
    @is(DataType.str).validate(required) @scope(fork)
    public action: "crop";

    @is(DataType.float).validate(required) @scope(fork)
    public start: number;

    @is(DataType.float).validate(required) @scope(fork)
    public end: number;
}

export class ForkOptions {
    @is().validate(required) @scope(fork)
    public description?: string;

    @is().validate(required) @scope(fork)
    public overwrite?: boolean;

    @is() @specify(() => Action)
    public actions: Action[];
}
