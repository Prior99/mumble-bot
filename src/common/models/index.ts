export { Recording } from "./recording";
export { User } from "./user";
export { Label } from "./label";
export { CachedAudio } from "./cached-audio";
export { RecordingLabelRelation } from "./recording-label-relation";
export { DatabaseSound } from "./database-sound";
export { Dialog } from "./dialog";
export { DialogPart } from "./dialog-part";
export { MumbleLink } from "./mumble-link";
export { Permission } from "./permission";
export { PermissionAssociation } from "./permission-association";
export { Token } from "./token";
export { MumbleUser } from "./mumble-user";
export { Channel } from "./channel";
export { PlaybackOptions } from "./playback-options";
export { ForkOptions } from "./fork-options";
export { Setting } from "./setting";
