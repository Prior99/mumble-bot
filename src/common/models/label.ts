import { Column, PrimaryGeneratedColumn, Entity, ManyToOne } from "typeorm";
import { is, scope, specify, uuid } from "hyrest";

import { world, createRecording } from "../scopes";

import { RecordingLabelRelation } from ".";

/**
 * A label with which the records can be tagged.
 */
@Entity()
export class Label {
    /**
     * Unique id of this record which is used as the mapping to the audio file.
     */
    @PrimaryGeneratedColumn("uuid")
    @scope(world, createRecording) @is().validate(uuid)
    public id?: string;

    /**
     * Name of this label.
     */
    @Column("varchar", { length: 100 })
    @is() @scope(world)
    public name?: string;

    /**
     * A list of all labels with which this record was tagged.
     */
    @ManyToOne(() => RecordingLabelRelation, recordingLabelRelation => recordingLabelRelation.label)
    @scope(world) @is() @specify(() => RecordingLabelRelation)
    public recordingLabelRelations?: RecordingLabelRelation[];
}
