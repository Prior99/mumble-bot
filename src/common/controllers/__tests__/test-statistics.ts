import { Connection } from "typeorm";
import {
    api,
    createUserWithTokenManually,
    startDb,
    stopDb,
    createTag,
    createUserManually,
    createSound,
    tagSound,
    rateSound,
} from "../../../test-utils";
import { Token, User, Sound } from "../../models";

describe("statistics controller", () => {
    beforeEach(startDb);
    afterEach(stopDb);

    let token: Token;

    beforeEach(async () => {
        const userAndToken = await createUserWithTokenManually({
            id: "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
        } as User);
        token = userAndToken.token;
    });

    async function insertMockData() {
        const tag1 = await createTag("Tag 1", token);
        const tag2 = await createTag("Tag 2", token);
        const user1 = await createUserManually({
            id: "2ce5e078-b52e-4d50-90cc-1703ec08ef2f",
            name: "User 1",
            email: "user1@example.com",
        } as User);
        const user2 = await createUserManually({
            id: "9e1d01d6-a203-4c86-9074-ee5e05a57dcd",
            name: "User 2",
            email: "user2@example.com",
        } as User);
        const sound1 = await createSound({
            description: "Some sound 1",
            user: user1,
            creator: user2,
            used: 100,
        } as Sound);
        await tsdi.get(Connection).getRepository(Sound).update(sound1.id, {
            created: new Date("2017-01-15T20:00:00"),
        });
        const sound2 = await createSound({
            description: "Some sound 2",
            user: user2,
            creator: user1,
            used: 900,
        } as Sound);
        await tsdi.get(Connection).getRepository(Sound).update(sound2.id, {
            created: new Date("2018-01-15T20:00:00"),
        });
        await tagSound(sound1, tag1, token);
        await tagSound(sound1, tag2, token);
        await rateSound(sound1.id, token.id, 2);
    }

    [
        "/statistics/overview",
        "/statistics/creations-per-user",
        "/statistics/playbacks-per-user",
        "/statistics/recordings-per-user",
        "/statistics/sounds-per-source",
        "/statistics/sounds-per-month",
    ].forEach(url => {
        describe(`GET ${url}`, () => {
            it("responds 401 with no valid token", async () => {
                const response = await api().get(url);
                expect(response.status).toBe(401);
            });

            it("responds 200 and the expected result with an empty database", async () => {
                const response = await api().get(url)
                    .set("authorization", `Bearer ${token.id}`);
                expect(response.status).toBe(200);
                expect(response.body.data).toMatchSnapshot();
            });

            it("responds 200 and the expected result with data in the database", async () => {
                await insertMockData();
                const response = await api().get(url)
                    .set("authorization", `Bearer ${token.id}`);
                expect(response.status).toBe(200);
                expect(response.body.data).toMatchSnapshot();
            });
        });
    });
});
