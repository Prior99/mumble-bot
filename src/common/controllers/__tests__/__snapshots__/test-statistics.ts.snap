// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statistics controller GET /statistics/creations-per-user responds 200 and the expected result with an empty database 1`] = `
Object {
  "creationsPerUser": Array [
    Object {
      "creations": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/creations-per-user responds 200 and the expected result with data in the database 1`] = `
Object {
  "creationsPerUser": Array [
    Object {
      "creations": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
    Object {
      "creations": 1,
      "user": Object {
        "id": "2ce5e078-b52e-4d50-90cc-1703ec08ef2f",
      },
    },
    Object {
      "creations": 1,
      "user": Object {
        "id": "9e1d01d6-a203-4c86-9074-ee5e05a57dcd",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/overview responds 200 and the expected result with an empty database 1`] = `
Object {
  "totalPlaybacks": 0,
  "totalSounds": 0,
  "totalTagged": 0,
  "totalTags": 0,
  "totalUnrated": 0,
  "totalUntagged": 0,
  "totalUsers": 1,
}
`;

exports[`statistics controller GET /statistics/overview responds 200 and the expected result with data in the database 1`] = `
Object {
  "oldestSound": "2017-01-15T20:00:00.000Z",
  "totalPlaybacks": 1000,
  "totalSounds": 2,
  "totalTagged": 2,
  "totalTags": 2,
  "totalUnrated": 1,
  "totalUntagged": 1,
  "totalUsers": 3,
}
`;

exports[`statistics controller GET /statistics/playbacks-per-user responds 200 and the expected result with an empty database 1`] = `
Object {
  "playbacksPerUser": Array [
    Object {
      "playbacks": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/playbacks-per-user responds 200 and the expected result with data in the database 1`] = `
Object {
  "playbacksPerUser": Array [
    Object {
      "playbacks": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
    Object {
      "playbacks": 100,
      "user": Object {
        "id": "2ce5e078-b52e-4d50-90cc-1703ec08ef2f",
      },
    },
    Object {
      "playbacks": 900,
      "user": Object {
        "id": "9e1d01d6-a203-4c86-9074-ee5e05a57dcd",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/recordings-per-user responds 200 and the expected result with an empty database 1`] = `
Object {
  "recordingsPerUser": Array [
    Object {
      "recordings": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/recordings-per-user responds 200 and the expected result with data in the database 1`] = `
Object {
  "recordingsPerUser": Array [
    Object {
      "recordings": 0,
      "user": Object {
        "id": "bc919169-6e7a-4ad8-a5c9-7e0b4f932b1e",
      },
    },
    Object {
      "recordings": 1,
      "user": Object {
        "id": "2ce5e078-b52e-4d50-90cc-1703ec08ef2f",
      },
    },
    Object {
      "recordings": 1,
      "user": Object {
        "id": "9e1d01d6-a203-4c86-9074-ee5e05a57dcd",
      },
    },
  ],
}
`;

exports[`statistics controller GET /statistics/sounds-per-month responds 200 and the expected result with an empty database 1`] = `
Object {
  "soundsPerMonth": Array [],
}
`;

exports[`statistics controller GET /statistics/sounds-per-month responds 200 and the expected result with data in the database 1`] = `
Object {
  "soundsPerMonth": Array [
    Object {
      "month": "2017-01-01T00:00:00.000Z",
      "sounds": 1,
    },
    Object {
      "month": "2018-01-01T00:00:00.000Z",
      "sounds": 1,
    },
  ],
}
`;

exports[`statistics controller GET /statistics/sounds-per-source responds 200 and the expected result with an empty database 1`] = `
Object {
  "soundsPerSource": Array [],
}
`;

exports[`statistics controller GET /statistics/sounds-per-source responds 200 and the expected result with data in the database 1`] = `
Object {
  "soundsPerSource": Array [
    Object {
      "sounds": 2,
      "source": "recording",
    },
  ],
}
`;
